### PR TITLE
cephadm: use shared spot for default container image names

### PR DESCRIPTION
**Pull request by adk3798**
Fixes: https://tracker.ceph.com/issues/68120

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/pybind/mgr/cephadm/module.py [hunk 1]
```diff
@@ -30,6 +30,7 @@
 import subprocess
 from prettytable import PrettyTable
 
+import ceph.cephadm.images as default_images
```
Comment (adk3798) [inline src/pybind/mgr/cephadm/module.py:33]: looks like there's still a docs thing to sort out
```
loading mgr module 'cephadm'...

Exception occurred:
  File "/home/jenkins-build/build/workspace/ceph-pr-docs/src/pybind/mgr/cephadm/module.py", line 33, in <module>
    import ceph.cephadm.images as default_images
ModuleNotFoundError: No module named 'ceph.cephadm'
```

**Review by rkachach — COMMENTED** [review]
I'm trying to review the code but I can't see where the images names are coming from.

**Note:** It would be helpful if you add some more details to the PR description about what would be the new source.

**Comment by adk3798** [comment]
> > I'm trying to review the code but I can't see where the images names are coming from.
> > **Note:** It would be helpful if you add some more details to the PR description about what would be the new source.
> 
> I forgot to add the new file while committing, oops. That probably explains the docs issue as well. One minute.

@rkachach the new file is present now

File: src/cephadm/build.py [hunk 2]
```diff
@@ -269,6 +269,9 @@ def _build(dest, src, config):
         mdir.mkdir(parents=True, exist_ok=True)
         (mdir / "__init__.py").touch(exist_ok=True)
         versioning_vars = config.cli_args.version_vars
+        shutil.copytree(
+            "../python-common/ceph", appdir / "ceph"
+        )
```
Comment (phlogistonjohn) [inline src/cephadm/build.py:274]: IMO this should be a separate commit that is clearly described in the commit message: as incorporating the ceph package from python-common into the cephadm binary. This is a significant enough change (in scope, not lines of code) that it deserves to be called out in the history.

**Review by phlogistonjohn — CHANGES_REQUESTED** [review]
I feel that this is trying to do too much in a single commit. I would break it up into N commits, roughly along these lines:
* including ceph package in cephadm binary (zipapp)
* fixing up tox files
* creating the new file
* replacing the old constants with the new file/import

HTH

File: src/python-common/ceph/cephadm/images.py
Comment (rkachach) [inline src/python-common/ceph/cephadm/images.py:1]: I have some doubts about this directory/package. We already have other files used by cephadm that are out of this directory (such as some of the files on the deployment directory). Are we planning to move everything which is specific to cephadm to this new package?

File: src/python-common/ceph/cephadm/images.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
+# Default container images -----------------------------------------------------
+DEFAULT_PROMETHEUS_IMAGE = 'quay.io/prometheus/prometheus:v2.51.0'
+DEFAULT_NODE_EXPORTER_IMAGE = 'quay.io/prometheus/node-exporter:v1.7.0'
+DEFAULT_NVMEOF_IMAGE = 'quay.io/ceph/nvmeof:1.2.17'
+DEFAULT_LOKI_IMAGE = 'quay.io/ceph/loki:3.0.0'
+DEFAULT_PROMTAIL_IMAGE = 'quay.io/ceph/promtail:3.0.0'
+DEFAULT_ALERT_MANAGER_IMAGE = 'quay.io/prometheus/alertmanager:v0.27.0'
```
Comment (adk3798) [inline src/python-common/ceph/cephadm/images.py:7]: I merged another PR that added some new import lines and didn't rebase this before testing, resulting in
```
2024-10-21T03:12:11.144 INFO:teuthology.orchestra.run.smithi133.stderr:Traceback (most recent call last):
2024-10-21T03:12:11.145 INFO:teuthology.orchestra.run.smithi133.stderr:  File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
2024-10-21T03:12:11.151 INFO:teuthology.orchestra.run.smithi162.stderr:Traceback (most recent call last):
2024-10-21T03:12:11.151 INFO:teuthology.orchestra.run.smithi162.stderr:  File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
2024-10-21T03:12:11.156 INFO:teuthology.orchestra.run.smithi133.stderr:    return _run_code(code, main_globals, None,
2024-10-21T03:12:11.156 INFO:teuthology.orchestra.run.smithi133.stderr:  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
2024-10-21T03:12:11.157 INFO:teuthology.orchestra.run.smithi133.stderr:    exec(code, run_globals)
2024-10-21T03:12:11.157 INFO:teuthology.orchestra.run.smithi133.stderr:  File "/tmp/tmpzfwsq3gn.cephadm.build/app/__main__.py", line 141, in <module>
2024-10-21T03:12:11.157 INFO:teuthology.orchestra.run.smithi133.stderr:  File "/tmp/tmpzfwsq3gn.cephadm.build/app/cephadmlib/runscripts.py", line 10, in <module>
2024-10-21T03:12:11.157 INFO:teuthology.orchestra.run.smithi133.stderr:  File "/tmp/tmpzfwsq3gn.cephadm.build/app/cephadmlib/container_types.py", line 11, in <module>
2024-10-21T03:12:11.157 INFO:teuthology.orchestra.run.smithi133.stderr:ImportError: cannot import name 'DEFAULT_ALERT_MANAGER_IMAGE' from 'cephadmlib.constants' (/home/ubuntu/cephtest/cephadm/cephadmlib/constants.pyc)
2024-10-21T03:12:11.159 INFO:teuthology.orchestra.run.smithi162.stderr:    return _run_code(code, main_globals, None,
2024-10-21T03:12:11.159 INFO:teuthology.orchestra.run.smithi162.stderr:  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
2024-10-21T03:12:11.160 INFO:teuthology.orchestra.run.smithi162.stderr:    exec(code, run_globals)
2024-10-21T03:12:11.160 INFO:teuthology.orchestra.run.smithi162.stderr:  File "/tmp/tmpzfwsq3gn.cephadm.build/app/__main__.py", line 141, in <module>
2024-10-21T03:12:11.160 INFO:teuthology.orchestra.run.smithi162.stderr:  File "/tmp/tmpzfwsq3gn.cephadm.build/app/cephadmlib/runscripts.py", line 10, in <module>
2024-10-21T03:12:11.160 INFO:teuthology.orchestra.run.smithi162.stderr:  File "/tmp/tmpzfwsq3gn.cephadm.build/app/cephadmlib/container_types.py", line 11, in <module>
2024-10-21T03:12:11.160 INFO:teuthology.orchestra.run.smithi162.stderr:ImportError: cannot import name 'DEFAULT_ALERT_MANAGER_IMAGE' from 'cephadmlib.constants' 
```
need to rebase this and fix it up

File: src/cephadm/build.py (hunk 2 above)
Comment (adk3798) [reply src/cephadm/build.py:274]: put this change in its own commit

**Comment by adk3798** [comment]
> I feel that this is trying to do too much in a single commit. I would break it up into N commits, roughly along these lines:
> 
> * including ceph package in cephadm binary (zipapp)
> * fixing up tox files
> * creating the new file
> * replacing the old constants with the new file/import
> 
> HTH

I've updated this to be 4 commits roughly following what you outlined here

File: src/python-common/ceph/cephadm/images.py
Comment (adk3798) [reply src/python-common/ceph/cephadm/images.py:1]: The stuff under ceph/deployment was intended to be used by more than just cephadm. At this point, it is in reality mostly just cephadm but the orchestrator setup does in theory allow other modules to become orchestrator backends and the stuff in python-common-deployment was more meant to be shared between that, while these default images were instead being shared between the cephadm binary and cephadm module. That said, I don't feel very strongly about it, so if we want it under ceph/deployment instead it's okay to me.

Comment (phlogistonjohn) [reply src/python-common/ceph/cephadm/images.py:1]: If this is meant to only be shared between the cephadm binary and the cephadm mgr module, I'm ok with this name. You might want to put that in a doc comment, say within __init__.py though.

File: src/cephadm/build.py (hunk 2 above)
Comment (rkachach) [reply src/cephadm/build.py:274]: @adk3798 as this location is for cephadm only shouldn't we create a cephadm subdirectory to make it clear instead of including the whole ceph dierctory?

File: src/cephadm/cephadmlib/container_types.py [hunk 4]
```diff
@@ -687,24 +667,46 @@ def get_mgr_images() -> dict:
...
+    mgr_images[
+        mgr_prefix + 'keepalived'
+    ] = default_images.DEFAULT_KEEPALIVED_IMAGE
+    mgr_images[mgr_prefix + 'nvmeof'] = default_images.DEFAULT_NVMEOF_IMAGE
+    mgr_images[
+        mgr_prefix + 'snmp_gateway'
+    ] = default_images.DEFAULT_SNMP_GATEWAY_IMAGE
+    mgr_images[
+        mgr_prefix + 'elasticsearch'
+    ] = default_images.DEFAULT_ELASTICSEARCH_IMAGE
```
Comment (rkachach) [inline src/cephadm/cephadmlib/container_types.py:694]: This could simplified/rewritten as: 

```
image_suffixes = [
    'prometheus',
    'alertmanager',
    'graphana',
    'loki',
    'promtail',
    'node_exporter',
    'haproxy',
    'keepalived',
    'nvmeof',
    'snmp_gateway',
    'elasticsearch',
]

for suffix in image_suffixes:
    key = mgr_prefix + suffix
    value = getattr(default_images, f'DEFAULT_{suffix.upper()}_IMAGE')
    mgr_images[key] = value

```

File: src/cephadm/build.py (hunk 2 above)
Comment (adk3798) [reply src/cephadm/build.py:274]: I think there's a chance in the future we'll end up wanting something from python-common/ceph/deployment and including the extra stuff in the zip doesn't have any downsides I'm aware of (I guess technically it takes up more space, but we're talking about kilobytes). That was my reasoning for just including the whole thing.

File: src/cephadm/cephadmlib/container_types.py (hunk 4 above)
Comment (adk3798) [reply src/cephadm/cephadmlib/container_types.py]: swapped to your second suggestion and tested it and it works nicely. It also helped discover a bug in the previous code as the config option name for grafana was incorrect (spelled "graphana") and the some of the SMB names differed as well.

```
[root@vm-00 ~]# diff old.txt new.txt 
3,4c3,4
< mgr/cephadm/container_image_alertmanager = quay.io/prometheus/alertmanager:v0.27.0
< mgr/cephadm/container_image_graphana = quay.io/ceph/grafana:10.4.8
---
> mgr/cephadm/container_image_node_exporter = quay.io/prometheus/node-exporter:v1.7.0
> mgr/cephadm/container_image_nvmeof = quay.io/ceph/nvmeof:1.2.17
7c7,8
< mgr/cephadm/container_image_node_exporter = quay.io/prometheus/node-exporter:v1.7.0
---
> mgr/cephadm/container_image_alert_manager = quay.io/prometheus/alertmanager:v0.27.0
> mgr/cephadm/container_image_grafana = quay.io/ceph/grafana:10.4.8
10d10
< mgr/cephadm/container_image_nvmeof = quay.io/ceph/nvmeof:1.2.17
15,17d14
< mgr/cephadm/container_image_jaeger_query = quay.io/jaegertracing/jaeger-query:1.29
< mgr/cephadm/container_image_smb = quay.io/samba.org/samba-server:devbuilds-centos-amd64
< mgr/cephadm/container_image_smbmetrics = quay.io/samba.org/samba-metrics:latest
19a17,19
> mgr/cephadm/container_image_jaeger_query = quay.io/jaegertracing/jaeger-query:1.29
> mgr/cephadm/container_image_samba = quay.io/samba.org/samba-server:devbuilds-centos-amd64
> mgr/cephadm/container_image_samba_metrics = quay.io/samba.org/samba-metrics:latest
```

Comment (adk3798) [reply src/cephadm/cephadmlib/container_types.py]: I did have to modify the alertmanager variable name for this to work as it had a "_" between alert and manager before, but that's fine

File: src/python-common/ceph/cephadm/images.py
Comment (adk3798) [reply src/python-common/ceph/cephadm/images.py:1]: added a comment in the __init__.py

File: src/cephadm/build.py (hunk 2 above)
Comment (rkachach) [reply src/cephadm/build.py:274]: @adk3798 sounds good, we just have to keep an eye on this directory in case somebody else put some big new module inside :+1:

**Review by rkachach — APPROVED** [review]
LGTM

File: src/cephadm/build.py (hunk 2 above)
Comment (phlogistonjohn) [reply src/cephadm/build.py:274]: Including python-common/ceph is a foundational change that allows cephadm binary and the other parts of ceph's python codebase to share code. That hypothetical module that we would want to exclude would have to be pretty huge for us to notice a problem IMO.

**Review by phlogistonjohn — APPROVED** [review]

**Comment by phlogistonjohn** [comment]
make check fails due to a minor but real linter error related to these changes

File: src/cephadm/build.py (hunk 2 above)
Comment (rkachach) [reply src/cephadm/build.py:274]: @phlogistonjohn yeah, probably we should setup something so any change on the ` python-common/ceph` would include the `orchesrtators` team as reviewers ... though I'm not sure if that would help or not .. just something to consider.

Comment (phlogistonjohn) [reply src/cephadm/build.py:274]: Fair enough. I have even wondered if a more general `python-sig` should exist, but for now that would be good enough. That doesn't need to be part of this PR however. :-)

Comment (phlogistonjohn) [reply src/cephadm/build.py:274]: https://github.com/ceph/ceph/pull/60546

File: src/cephadm/cephadmlib/container_types.py (hunk 4 above)
Comment (adk3798) [reply src/cephadm/cephadmlib/container_types.py]: We're planning some further work to have the config setting for the images be specified in this same images.py file in the future which I think will cover that.

**Comment by adk3798** [comment]
```
99% tests passed, 1 tests failed out of 302

Total Test time (real) = 2272.86 sec

The following tests FAILED:
	 33 - run-rbd-unit-tests-109.sh (Failed)
```

**Comment by adk3798** [comment]
jenkins test make check

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2024-10-30_03:37:52-orch:cephadm-wip-adk-testing-2024-10-29-1832-distro-default-smithi/

Failures:
- 3x SMB ctdb test stray daemon / SELinux denial, known issue
- 2x failures on new tests for mgmt-gateway, caused by PR adding those new tests
- 3x mds_upgrade_sequence failures, known issue
- 1x staggered upgrade with agent failure, known issue

Dead Jobs:
- 1x `Error reimaging machines`, infra/teuthology issue
- 1x rgw-ingress test timeout, known issue
- 1x test_iscsi_container timeout. I've seen this before but thought it had been fixed. Will need to investigate.

Results should only block merging for PRs related to iscsi or mgmt-gateqay/oauth2-proxy. Otherwise, run was okay.
